File: cibcc/Go/step3/cibcc.go
package main

import (
	"os"
	"fmt"
)

type TokenKind	int

const (
	TK_RESERVED TokenKind = iota
	TK_NUM
	TK_EOF
)

type Token struct {
	kind TokenKind
	next *Token
	val int
	str []rune
}

var token *Token


func isSpace(c rune) bool {
	return c == ' ' || c == '\t' || c == '\n' || c ==  '\r'
}

func isDigit(c rune) bool {
	return '0' <= c && c <= '9'
}

func isReserved(c rune) bool {
	switch c {
		case '+':	fallthrough
		case '-':	return true
		default:	return false
	}
}

func strtoi(p *[]rune) int {
	s := *p
	c :=s[0]
	s = s[1:]

	acc := 0
	for {
		k := int(c - '0')
		acc *= 10
		acc += k
		if len(s) == 0 || !isDigit(s[0]) {
			break
		}
		c = s[0]
		s = s[1:]
	}
	*p = s
	return acc
}

func new_Token(kind TokenKind, cur *Token, str []rune) *Token {
	p := &Token{kind: kind, str: str}
	cur.next = p
	return p
}

func tokenize(p []rune) *Token {
	var head Token
	head.next = nil
	cur := &head

	for len(p) > 0 {
		if isSpace(p[0]) {
			p = p[1:]
			continue
		}
		if isReserved(p[0]) {
			cur = new_Token(TK_RESERVED, cur, p) 
			p = p[1:]
			continue
		}
		if isDigit(p[0]) {
			cur = new_Token(TK_NUM, cur, p)
			v := strtoi(&p)
			cur.val = v
			continue
		}
	}
	new_Token(TK_EOF, cur, p)
	return head.next
}

func consume(op rune) bool {
	if token.kind != TK_RESERVED || token.str[0] != op {
		return false
	}
	token = token.next
	return true
}

func expect(op rune) {
	if token.kind != TK_RESERVED || token.str[0] != op {
		fmt.Println("not expected code")
	}
	token = token.next
}

func expect_number() int {
	v := token.val
	token = token.next
	return v
}

func at_eof() bool {
	return token.kind == TK_EOF
}


func main() {
	if len(os.Args) != 2 {
		fmt.Println("引数の個数が正しくありません")
	}

	token = tokenize([]rune(os.Args[1]))

	fmt.Println(".intel_syntax noprefix")
	fmt.Println(".global main")
	fmt.Println("main:")

	fmt.Println("  mov rax, ", expect_number())

	for !at_eof() {
		if consume('+') {
			fmt.Printf("  add rax, %d\n", expect_number())
		} else if consume('-') {
			fmt.Printf("  sub rax, %d\n", expect_number())
		}
	}
	fmt.Println("  ret")
}

File: cibcc/c++/step3/cibcc.cpp
#include <iostream>
#include <memory>
#include <vector>

enum TokenKind {
	TK_RESERVED,
	TK_NUM,
	TK_EOF,
};

struct Token {
	TokenKind kind;
	Token *next;
	int val;
	char *str;

	Token() : kind(TK_EOF), next(nullptr), val(0), str(nullptr) {}

	Token(TokenKind t, char *s)
		: kind(t), next(nullptr), val(0), str(s) {}
};

Token *token;

Token *tokenize(char *p);

Token *tokenize(char *p) {
	Token head;
	Token *cur = &head;

	while (*p) {
		if (isspace(*p)) {
			p++;
			continue;
		}

		if (*p == '+' || *p == '-') {
			cur->next = new Token(TK_RESERVED, p);
			cur = cur->next;
			p++;
			continue;
		}

		if (isdigit(*p)) {
			cur->next = new Token(TK_NUM, p);
			cur = cur->next;
			cur->val = strtol(p, &p, 10);
			continue;
		}
		std::cerr << "invalid token\n";
	}

	cur->next = new Token(TK_EOF, p);
	return head.next;
}

int main(int argc, char* argv[]) {
	if (argc != 2) {
		std::cout << "引数の個数が正しくありません\n";
	return 1;
	}

	token = tokenize(argv[1]);

	std::cout << ".intel_syntax noprefix\n";
	std::cout << ".global main\n";
	std::cout << "main:\n";

	if (token->kind != TK_NUM) {
		std::cout << "数ではありません\n";
		return 1;
	}

	std::cout << "  mov rax, " << token->val << "\n";
	token = token->next;

	while (token->kind != TK_EOF) {
		if (token->kind == TK_RESERVED && *token->str == '+') {
			token = token->next;
			if (token->kind != TK_NUM) {
				std::cout << "数ではありません\n";
				return 1;
			}
			std::cout << "  add rax, " << token->val << "\n";
			token = token->next;
			continue;
		}
		if (token->kind == TK_RESERVED && *token->str == '-') {
			token = token->next;
			if (token->kind != TK_NUM) {
				std::cout << "数ではありません\n";
				return 1;
			}
			std::cout << "  sub rax, " << token->val << "\n";
			token = token->next;
			continue;
		}

		std::cout << "予期しないトークンです\n";
		return 1;
	}
	std::cout << "  ret\n";
	return 0;
}

